Source/Core/VideoCommon/ImGuiManager.cpp

```diff
@@ -0,0 +1,380 @@
+  ImGui::NewFrame();
+  return true;
+}
+
+void ImGuiManager::Shutdown()
+{
+  ImGui::EndFrame();
+  ImGui::DestroyContext();
+  m_imgui_pipeline.reset();
+  m_imgui_vertex_format.reset();
+  m_imgui_textures.clear();
+}
+
+void ImGuiManager::Render()
+{
+  ImGui::Render();
+
+  ImDrawData* draw_data = ImGui::GetDrawData();
+  if (!draw_data || g_renderer->IsHeadless())
+    return;
+
+  ImGuiIO& io = ImGui::GetIO();
+  g_renderer->SetViewport(0.0f, 0.0f, io.DisplaySize.x, io.DisplaySize.y, 0.0f, 1.0f);
+
+  // Uniform buffer for draws.
+  struct ImGuiUbo
+  {
+    float u_rcp_viewport_size_mul2[2];
+    float padding[2];
+  };
+  ImGuiUbo ubo = {{1.0f / io.DisplaySize.x * 2.0f, 1.0f / io.DisplaySize.y * 2.0f}};
+
+  // Set up common state for drawing.
+  g_renderer->SetPipeline(m_imgui_pipeline.get());
+  g_renderer->SetSamplerState(0, RenderState::GetPointSamplerState());
+  g_vertex_manager->UploadUtilityUniforms(&ubo, sizeof(ubo));
+
+  for (int i = 0; i < draw_data->CmdListsCount; i++)
+  {
+    const ImDrawList* cmdlist = draw_data->CmdLists[i];
+    if (cmdlist->VtxBuffer.empty() || cmdlist->IdxBuffer.empty())
+      return;
+
+    u32 base_vertex, base_index;
+    g_vertex_manager->UploadUtilityVertices(cmdlist->VtxBuffer.Data, sizeof(ImDrawVert),
+                                            cmdlist->VtxBuffer.Size, cmdlist->IdxBuffer.Data,
+                                            cmdlist->IdxBuffer.Size, &base_vertex, &base_index);
+
+    for (const ImDrawCmd& cmd : cmdlist->CmdBuffer)
+    {
+      if (cmd.UserCallback)
+      {
+        cmd.UserCallback(cmdlist, &cmd);
+        continue;
+      }
+
+      g_renderer->SetScissorRect(g_renderer->ConvertFramebufferRectangle(
+          MathUtil::Rectangle<int>(
+              static_cast<int>(cmd.ClipRect.x), static_cast<int>(cmd.ClipRect.y),
+              static_cast<int>(cmd.ClipRect.z), static_cast<int>(cmd.ClipRect.w)),
+          g_renderer->GetCurrentFramebuffer()));
+      g_renderer->SetTexture(0, reinterpret_cast<const AbstractTexture*>(cmd.TextureId));
+      g_renderer->DrawIndexed(base_index, cmd.ElemCount, base_vertex);
+      base_index += cmd.ElemCount;
+    }
+  }
+}
+
+std::unique_lock<std::mutex> ImGuiManager::GetStateLock()
+{
+  return std::unique_lock<std::mutex>(m_state_mutex);
+}
+
+void ImGuiManager::EndFrame()
+{
+  std::unique_lock<std::mutex> imgui_lock(m_state_mutex);
+
+  const u64 current_time_us = Common::Timer::GetTimeUs();
+  const u64 time_diff_us = current_time_us - m_last_frame_time;
+  const float time_diff_secs = static_cast<float>(time_diff_us / 1000000.0);
+  m_last_frame_time = current_time_us;
+  ImGui::NewFrame();
+
+  // Update I/O with window dimensions.
+  ImGuiIO& io = ImGui::GetIO();
+  io.DisplaySize.x = static_cast<float>(g_renderer->GetBackbufferWidth());
+  io.DisplaySize.y = static_cast<float>(g_renderer->GetBackbufferHeight());
+  io.DeltaTime = time_diff_secs;
+}
+
+void ImGuiManager::Draw()
```

Comment:
Would there be any benefit in writing this in a way that other places can register themselves to be rendered here (like, `for (auto* drawable : m_drawables) drawable->Draw()`), instead of a long `if (drawA) renderA();` chain?

Reply:
I like this idea, if it's done, it'd be nice to have a way to unregister as well (thinking of pause screen)

Reply:
How about a bitmask of states where it should be displayed?

Reply:
You could just pass that in, if its the same method for all states. Or just have methods for individual states (not sure which one makes more sense).
Having a Register/Unregister mechanism here would make it a lot nicer to plug new things into the ImGUI UI without having to dig around this manager and related files.

Reply:
Yeah, sorry, I meant using a bitmask in combination with the callback registration idea. This would save having to check which state the UI was in to display a window or not.